### ISSUE TITLE
Recommend `match` instead of `.map().unwrap_or_else()`

### ISSUE THREAD

**Issue by gendx** [rust-lang/rust-clippy#17470]
### What it does

Warn against `x.map(...).unwrap_or_else(...)` on an `Option`/`Result`, and recommend a `match` instead. Similar to https://github.com/rust-lang/rust-clippy/issues/17469.

### Advantage

- More readable.
- Clearer to the reader and to the compiler that the `match` is exhaustive.

### Drawbacks

None that I can think of, other than style and personal taste.

### Example

```rust
opt.map(f).unwrap_or_else(g)
```

Could be written as:

```rust
match opt {
    Some(x) => f(x),
    None => g(),
}
```

And for a `Result`:

```rust
match res {
    Ok(x) => f(x),
    Err(e) => g(e),
}
```

**Comment by qdot3** [comment]
`map_or_else()` is also a good option. I prefer `match` when the closures are long or involve blocks, and `map_or_else` when they are short and simple.

**Comment by qdot3** [comment]
In a long method chain (e.g. `a().b().c().map().unwrap_or_else()`), `unwrap_or_else` often reads better because it keeps the chain consistent and avoids breaking it with a `match`.

**Comment by gendx** [comment]
There are definitely stylistic choices here as to which style is more readable so I wouldn't put this lint into the default denylist unless there's broad consensus. However, having it available somewhere in the `pedantic` list to allow enabling it on a per-project basis would be useful.

**Comment by qdot3** [comment]
IMO, the difficulty with `map().unwrap_or_else()` is that the branching becomes implicit and split across the chain, while `match` and `map_or_else` keep both branches explicit and localized.

On the other hand, a chain like `map().map().map().unwrap_or_else()` has a certain conciseness: every step clearly transforms the `Some/Ok` value, and the control flow stays linear without branching.

Of course, it’s a matter of personal preference.
